source/agora/consensus/data/Block.d

```diff
@@ -156,6 +157,30 @@ public struct Block
         hashPart(this.header, dg);
     }
 
+    /***************************************************************************
+
+        Create the block signature for the given keypair
+
+        This signature will be combined with other validator's signatures
+        using Schnorr multisig.
+
+        Params:
+            block = the block to sign
+            secret_key = node's secret
+            offset = enrollment cycle offset
+
+    ***************************************************************************/
+    public Signature createBlockSignature (in Scalar secret_key, ulong offset = 0) const @safe nothrow
+    {
+        // challenge = Hash(block) to Scalar
+        const Scalar challenge = this.hashFull();
+        // rc = r used in signing the commitment
+        const Scalar rc = Scalar(hashMulti(secret_key, "consensus.signature.noise", offset));
+        const Scalar r = rc + challenge; // make it unique each challenge
+        const Point R = r.toPoint();
+        return sign(secret_key, R, r, challenge);
+    }
+
```

Comment:
I think it should be in Header, no ?

Reply:
Good point, I'll move it.

Reply:
Although the header is updated, it is the `block` that is hashed and then signed. Currently the block hash is that of the header. Not sure if logically it should move though. What do you think?

Reply:
ok I moved the function with the assumption that we will always use the `computeHash` of the `BlockHeader` for signing blocks.